### PR TITLE
Speedup codec NONE

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Avoid unnecessary data copy when using codec `NONE`. Please note that codec `NONE` is mostly useless - it's recommended to always use compression (`LZ4` is by default). Despite the common belief, disabling compression may not improve performance (the opposite effect is possible). The `NONE` codec is useful in some cases: - when data is uncompressable; - for synthetic benchmarks.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Two times performance improvement:

```
CREATE TABLE test.hits_none (Title String CODEC(NONE)) ENGINE = MergeTree ORDER BY tuple();
INSERT INTO test.hits_none SELECT Title FROM test.hits;
OPTIMIZE TABLE test.hits_none FINAL;
```

```
clickhouse-benchmark --query "SELECT sum(length(Title)) FROM test.hits_none"
```

Before: 
```
localhost:9000, queries 210, QPS: 7.624, RPS: 67652204.332, MiB/s: 5576.078, result RPS: 7.624, result MiB/s: 0.000.

0.000%          0.114 sec.
10.000%         0.122 sec.
20.000%         0.124 sec.
30.000%         0.126 sec.
40.000%         0.128 sec.
50.000%         0.130 sec.
```

After:
```
localhost:9000, queries 248, QPS: 15.142, RPS: 134366240.938, MiB/s: 11074.830, result RPS: 15.142, result MiB/s: 0.000.

0.000%          0.055 sec.
10.000%         0.059 sec.
20.000%         0.061 sec.
30.000%         0.063 sec.
40.000%         0.065 sec.
50.000%         0.066 sec.
```

PS. Compressed hits table is faster on my machine:
```
localhost:9000, queries 597, QPS: 21.958, RPS: 194853523.858, MiB/s: 16060.356, result RPS: 21.958, result MiB/s: 0.000.

0.000%          0.038 sec.
10.000%         0.041 sec.
20.000%         0.042 sec.
30.000%         0.043 sec.
40.000%         0.044 sec.
50.000%         0.045 sec.
```

**Comment by alexey-milovidov** [comment]
Something is wrong with mmap IO: 
```
clickhouse-benchmark --min_bytes_to_use_mmap_io 1 --max_threads 1 --query "SELECT sum(length(Title)) FROM test.hits_none"
```

**Comment by alexey-milovidov** [comment]
TODO:
- add perf test;
- add scenario with MMap IO to test.

**Review by kitaisreal — APPROVED** [review]

File: src/Compression/CompressedReadBufferBase.cpp [hunk 1]
```diff
@@ -210,11 +210,38 @@ void CompressedReadBufferBase::decompress(char * to, size_t size_decompressed, s
...
+    readHeaderAndGetCodec(compressed_buffer, size_decompressed, codec, allow_different_codecs);
     codec->decompress(compressed_buffer, size_compressed_without_checksum, to);
 }
 
 
+void CompressedReadBufferBase::decompress(BufferBase::Buffer & to, size_t size_decompressed, size_t size_compressed_without_checksum)
+{
+    readHeaderAndGetCodec(compressed_buffer, size_decompressed, codec, allow_different_codecs);
+
+    if (codec->isNone())
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Compression/CompressedReadBufferBase.cpp:227]: It's unlikely that unlikely will make sense here.

Likely/unlikely is for fine tunes in machine code layout and here we have at least virtual function call.

File: src/Compression/CompressedReadBufferBase.cpp [hunk 2]
```diff
@@ -210,11 +210,38 @@ void CompressedReadBufferBase::decompress(char * to, size_t size_decompressed, s
...
+        /// Shortcut for NONE codec to avoid extra memcpy.
+        /// We doing it by changing the buffer `to` to point to existing uncompressed data.
+
+        UInt8 header_size = ICompressionCodec::getHeaderSize();
+        if (size_compressed_without_checksum < header_size)
+            throw Exception(ErrorCodes::CORRUPTED_DATA,
+                "Can't decompress data: the compressed data size ({}), this should include header size) is less than the header size ({})",
+                    size_compressed_without_checksum, size_t(header_size));
+
+        to = BufferBase::Buffer(compressed_buffer + header_size, compressed_buffer + size_compressed_without_checksum);
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Compression/CompressedReadBufferBase.cpp:238]: I tried to do it but there is no `memory` member in `CompressedReadBufferBase`.

File: src/Compression/CompressedReadBufferBase.cpp [hunk 3]
```diff
@@ -210,11 +210,38 @@ void CompressedReadBufferBase::decompress(char * to, size_t size_decompressed, s
...
+    if (codec->isNone())
+    {
+        /// Shortcut for NONE codec to avoid extra memcpy.
+        /// We doing it by changing the buffer `to` to point to existing uncompressed data.
+
+        UInt8 header_size = ICompressionCodec::getHeaderSize();
+        if (size_compressed_without_checksum < header_size)
+            throw Exception(ErrorCodes::CORRUPTED_DATA,
+                "Can't decompress data: the compressed data size ({}), this should include header size) is less than the header size ({})",
+                    size_compressed_without_checksum, size_t(header_size));
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Compression/CompressedReadBufferBase.cpp]: Ok.

**Comment by alexey-milovidov** [comment]
Performance test has shown the improvement.

**Pull request merged**